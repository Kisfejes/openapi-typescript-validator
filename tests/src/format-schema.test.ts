import path from "path";
import fs from "fs";
import { generate } from "openapi-typescript-validator";

describe("format-schema - compile based", () => {
  const name = "format";
  const generatedDir = path.join(__dirname, "../generated", name);
  const schemaDir = path.join(__dirname, "../schemas");

  beforeAll(async () => {
    if (fs.existsSync(generatedDir))
      fs.rmdirSync(generatedDir, { recursive: true });
    await generate({
      schemaFile: path.join(schemaDir, "format-schema.js"),
      schemaType: "custom",
      name,
      directory: generatedDir,
      addFormats: true,
    });
  });

  test("files should match", () => {
    const files = fs.readdirSync(generatedDir);
    expect(files).toEqual([
      "decoders.ts",
      "helpers.ts",
      "meta.ts",
      "models.ts",
      "schema.json",
    ]);
  });

  test("schema should match", () => {
    const file = fs.readFileSync(
      path.join(generatedDir, `schema.json`),
      "utf8"
    );
    expect(file).not.toBeUndefined();
    expect(file).toMatchSnapshot();
  });

  test("decoders should match", () => {
    const file = fs.readFileSync(
      path.join(generatedDir, `decoders.ts`),
      "utf8"
    );
    expect(file).not.toBeUndefined();
    expect(file).toMatchSnapshot();
  });

  test("models should match", () => {
    const file = fs.readFileSync(path.join(generatedDir, `models.ts`), "utf8");
    expect(file).not.toBeUndefined();
    expect(file).toMatchSnapshot();
  });
});
