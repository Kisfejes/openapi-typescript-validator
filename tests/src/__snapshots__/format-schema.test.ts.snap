// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format-schema - compile based decoders should match 1`] = `
"/* eslint-disable */

import Ajv from \\"ajv\\";
import addFormats from \\"ajv-formats\\";
import { validateJson, Decoder } from \\"./helpers\\";
import { User } from \\"./models\\";
import jsonSchema from \\"./schema.json\\";

const ajv = new Ajv({ strict: false });
ajv.compile(jsonSchema);
addFormats(ajv);

// Decoders
export const UserDecoder: Decoder<User> = {
  definitionName: \\"User\\",
  schemaRef: \\"#/definitions/User\\",

  decode(json: unknown): User {
    const schema = ajv.getSchema(UserDecoder.schemaRef);
    if (!schema) {
      throw new Error(\`Schema \${UserDecoder.definitionName} not found\`);
    }
    return validateJson(json, schema, UserDecoder.definitionName);
  },
};
"
`;

exports[`format-schema - compile based models should match 1`] = `
"/* eslint-disable */
/* tslint:disable */
/**
 * This file was automatically generated by json-schema-to-typescript.
 * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
 * and run json-schema-to-typescript to regenerate this file.
 */

export interface User {
  createdAt: string;
}
"
`;

exports[`format-schema - compile based schema should match 1`] = `
"{
  \\"type\\": \\"object\\",
  \\"title\\": \\"Schema\\",
  \\"definitions\\": {
    \\"User\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"createdAt\\": {
          \\"type\\": \\"string\\",
          \\"format\\": \\"date-time\\"
        }
      },
      \\"required\\": [
        \\"createdAt\\"
      ]
    }
  },
  \\"properties\\": {
    \\"User\\": {
      \\"$ref\\": \\"#/definitions/User\\"
    }
  }
}"
`;
